technical-docs/2021-05-remote-restore-redesign.md

```diff
@@ -0,0 +1,122 @@
+
+## CassandraBackup CRD changes
+CassandraBackup resources shouldn’t store the CassandraDatacenter spec anymore but rather the topology object generated by Medusa.  
+This object will contain the list of live nodes at the time of the backup along with their token assignments.
+Sample topology object:
+```
+{"ip-179-31-16-38.us-west-2.compute.internal": {"tokens": [-1051417974279770378, -112095142485141927, 9217379683620055011, 939597040143840721], "is_up": true}, 
+ "ip-179-31-23-54.us-west-2.compute.internal": {"tokens": [-1080182033668782299, -1156652990105327691, 9134535641516355167, 9216775765860996507], "is_up": true}, 
+ "ip-179-31-30-139.us-west-2.compute.internal": {"tokens": [-1018393144648419592, -1342532530495729638, 8846603596658721956, 8849342512812637165], "is_up": true}}
+```
+
+The CRD should also contain information that should be overridable such as the prefix, which is controlled by the namespace name. 
+
+## Medusa changes
+The tokenmap generated by Medusa will need to be augmented to provide the placement of nodes in the cluster: datacenter and rack.
+Currently Medusa is not dc/rack aware and it can probably be fairly surprising to users. It was built to backup a single DC per cluster and didn’t account for racks either.  
+Restoring from a DC/rack aware cluster to another one imposes some constraints on the source/target nodes mapping, which are currently impossible to deal with safely in a remote restore: 
+
+When using NetworkTopologyStrategy, the placement of replicas is dependent on the DC and rack layout.
+In this context, the replica for a token range will be placed on the node that owns the next token range in the next rack.
+We can't restore any node to any other node as we may end up putting several replicas in the same rack, and the nodes we'd restore to would possibly end up not containing data they're supposed be replica for.
+We can be rigid about the mapping and expect the same dc and rack names exclusively, or do a 1:1 mapping between DCs/Racks despite having different names (but expect the same number of both).
+
+
+## Restore Orchestration
+
+1. Get backup definition from storage backend
+![](img/K8ssandra-new-restore-design-1.jpg)  
+The restore operation will assume the CRDs have been synced with the remote storage backend. Backup topology (hostnames and tokens) will be retrieved from the new CRD format.
+
+
+2. Set the env variables on the restore containers
+![](img/K8ssandra-new-restore-design-2.jpg)  
+The backup definition will be used to map the backup hosts with the restore hosts and perform checks on whether or not the restore is possible:
+
+  - Same number of DCs
+  - Same number of nodes per DC
+  - Same rack distribution per DC
+  
+  Nodes will be mapped 1:1 between the source and destination topologies (sorted alphabetically). If the mapping fails to associate nodes 1:1, the restore operation must fail.
```

Comment:
Ok that answers my question above :-)

Reply:
Also, do we need to check for same token ranges? Or this doesn't matter? If origin and target nodes have different ranges, I wonder if target could end up containing data that don't belong to it.

Reply:
Medusa will enforce the tokens to match the original topology. We need to make sure the backup node and restore node are placed the same in the NetworkTopologyStrategy context, otherwise we'll have some replicated tokens that won't be accessible indeed.
So given we have the same number of DCs and racks (which is a prereq), a backup from a node in the first DC and the second rack must be restored on a node in the first DC and the second rack in the target cluster.